theano/gpuarray/opt.py

```diff
@@ -165,25 +169,36 @@ def f(local_opt):
 
 def safe_to_gpu(x, ctx_name):
     if isinstance(x.type, tensor.TensorType):
-        return gpu_from_host(ctx_name)(x)
+        return GpuFromHost(ctx_name)(x)
     else:
         return x
 
 
 def safe_to_cpu(x):
     if isinstance(x.type, GpuArrayType):
-        return host_from_gpu(x)
+        return x.transfer('cpu')
     else:
         return x
 
 
+def register_topo(*tags, **kwargs):
+    def f(local_opt):
+        name = (kwargs and kwargs.pop('name')) or local_opt.__name__
+        name = name + "_graph"
+        gpu_topo.register(name, local_opt,
+                          10, 'fast_run', 'gpuarray', *tags)
+        return local_opt
+    return f
```

Comment:
I remember `Alloc` was a special case and the nodes containing this OP were not move to the GPU. That is why I had done a special handling of GPUAlloc.